### PR TITLE
[CONTP-388] switch to composite store and use composite entity id in tagger component

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

This PR switches the tagger to the composite implementation which includes the following:
- converting entity id from string (of the form `prefix://id`) to a struct of prefix and id.
- storing entity tags in a 2 layer map in the form of `map[prefix]map[id]Object`

### Motivation

Remove default tagger implementation and use the new composite implementation which provides better performance for filtering and CLC runners.

### Describe how to test/QA your changes

No need for QA, E2E and unit tests should be enough.

### Additional Notes

Testing with single-performance-machine regression detector shows that no regression is introduced neither in cpu nor in memory performance.

### CODE REVIEW

File: comp/core/tagger/component.go [hunk 1]
```diff
@@ -39,13 +39,13 @@ type Component interface {
 	ReplayTagger() ReplayTagger
 	GetTaggerTelemetryStore() *telemetry.Store
 	Tag(entityID string, cardinality types.TagCardinality) ([]string, error)
```
Comment (adel121) [inline comp/core/tagger/component.go:41]: I didn't update the signature of the `Tag` method to avoid breaking rtloader. 

In a separate PR, I will add a new method to the tagger component (e.g. `TagWithEntityID`) which receives `entityID` as a struct, and will update calls to the tagger to use the new method, without breaking python integrations.

**Review by vickenty — APPROVED** [review]

**Review by duncanista — APPROVED** [review]
LGTM on the Serverless side

**Review by GustavoCaso — APPROVED** [review]
LGTM 🎉

**Review by ogaca-dd — COMMENTED** [review]
Is it safe to rely only on unit tests and e2e tests?
Also the [previous PR](https://github.com/DataDog/datadog-agent/pull/28645) also relied only on unit tests and e2e tests.

**Comment by adel121** [comment]
> Is it safe to rely only on unit tests and e2e tests? Also the [previous PR](https://github.com/DataDog/datadog-agent/pull/28645) also relied only on unit tests and e2e tests.

@ogaca-dd 

The main change here is done to the tagger and this change should not change the behaviour of the tagger for the external components, it is more like a performance optimisation change.

Given this, there isn't really a specific thing that we can manually QA here. To validate this change the tagger should continue working as it worked before (collect, store and expose tags correctly). Given the many usecases of this functionality, E2E tests and monitors make a better job on doing a QA check. If tags were missing, or not exposed correctly, at least 1 test should fail.

Additionally, I enabled the composite store implementation on gizmo a few days ago to see if everything looks good.

**Review by CelianR — APPROVED** [review]

**Review by KevinFairise2 — APPROVED** [review]

File: pkg/collector/corechecks/containers/kubelet/provider/cadvisor/provider_test.go [hunk 2]
```diff
@@ -107,7 +108,8 @@ func (suite *ProviderTestSuite) SetupTest() {
 	fakeTagger := taggerimpl.SetupFakeTagger(suite.T())
 	defer fakeTagger.ResetTagger()
 	for entity, tags := range commontesting.CommonTags {
-		fakeTagger.SetTags(entity, "foo", tags, nil, nil, nil)
+		entityID, _ := taggertypes.NewEntityIDFromString(entity)
```
Comment (L3n41c) [inline pkg/collector/corechecks/containers/kubelet/provider/cadvisor/provider_test.go:111]: Would it make sense to change the type and the definition of `commontesting.CommonTags` to change the type of keys from `string` to `EntityID` so that we don’t have to call `NewEntityIDFromString` here?

This is a test, so we don’t care about performances. But the goal of this PR is to reduce the usage of structured `EntityID` to and from `string` conversions and the other way round.
And I fear that any piece of code, including tests, can be used as a source of inspiration for developers writing pieces of code invoked in the hot path.
That’s why I think it would make sense to get rid of `NewEntityIDFromString` invocations in tests as well.

File: pkg/collector/corechecks/containers/kubelet/provider/probe/provider_test.go [hunk 3]
```diff
@@ -274,7 +275,8 @@ func TestProvider_Provide(t *testing.T) {
 			fakeTagger := taggerimpl.SetupFakeTagger(t)
 			defer fakeTagger.ResetTagger()
 			for entity, tags := range probeTags {
-				fakeTagger.SetTags(entity, "foo", tags, nil, nil, nil)
+				entityID, _ := taggertypes.NewEntityIDFromString(entity)
```
Comment (L3n41c) [inline pkg/collector/corechecks/containers/kubelet/provider/probe/provider_test.go:278]: Here also, we may want to change the type of `probeTags` above in this file from `map[string][]string` to `map[EntityID][]string` to be able to get rid of `NewEntityIDFromString` invocations here.

File: pkg/collector/corechecks/containers/kubelet/provider/summary/provider_test.go [hunk 4]
```diff
@@ -336,7 +337,8 @@ func TestProvider_Provide(t *testing.T) {
 			fakeTagger := taggerimpl.SetupFakeTagger(t)
 			defer fakeTagger.ResetTagger()
 			for entity, tags := range entityTags {
-				fakeTagger.SetTags(entity, "foo", tags, nil, nil, nil)
+				entityID, _ := taggertypes.NewEntityIDFromString(entity)
```
Comment (L3n41c) [inline pkg/collector/corechecks/containers/kubelet/provider/summary/provider_test.go:340]: Same for `entityTags` here.

File: pkg/collector/corechecks/containers/kubelet/provider/cadvisor/provider_test.go (hunk 2 above)
Comment (adel121) [reply pkg/collector/corechecks/containers/kubelet/provider/cadvisor/provider_test.go:111]: My intent was to keep unit tests easy to write by supplying plain strings in each testcase. 

But I also agree it might serve as an inspiration for other developers to use `NewEntityIDFromString`, will update this. 

Thanks 👍

**Review by daniel-taf — APPROVED** [review]

File: comp/core/tagger/taggerimpl/generic_store/store_test.go [hunk 5]
```diff
@@ -11,206 +11,125 @@ import (
...
 
-	// default store
-	test(t, false)
-
-	// composite store
-	test(t, true)
+	// list should return empty
+	list = store.ListObjects(filter)
+	expectedListing := []any{"prefix1://id1", "prefix2://id2"}
+	assert.ElementsMatch(t, expectedListing, list)
```
[the remark this answers is not in the document]
Comment (adel121) [reply comp/core/tagger/taggerimpl/generic_store/store_test.go:111]: Because we are running the callback function only on the items that match the filter that we created [here](https://github.com/DataDog/datadog-agent/blob/eabaac24351503b13e1e4dc0406aaaaa457f124c/comp/core/tagger/taggerimpl/generic_store/store_test.go#L92-L95):

```
	// build some filter
	fb := types.NewFilterBuilder()
	fb.Include(types.EntityIDPrefix("prefix1"), types.EntityIDPrefix("prefix2"))
	filter := fb.Build(types.HighCardinality)
```

**Review by liustanley — APPROVED** [review]

**Comment by ogaca-dd** [comment]
> Given the many usecases of this functionality, E2E tests and monitors make a better job on doing a QA check. If tags were missing, or not exposed correctly, at least 1 test should fail.
Additionally, I enabled the composite store implementation on gizmo a few days ago to see if everything looks good.

@adel121 I think this is hard now to tell if a feature is covered or not by an e2e tests?
Anyway, your tests on gizmo seems to be enough (it should probably be added to the QA section).

**Comment by adel121** [comment]
@ogaca-dd 

> I think this is hard now to tell if a feature is covered or not by an e2e tests?

Yes I agree with you. 

But this PR is not about introducing a feature. 
It is rather an optimisation done to a middleware component (i.e. an internal component that serves other components).
E2E tests should test public user-facing features, and not internals. 

When modifying a central internal component, the ideal scenario is that every user-facing feature that uses this internal component has an E2E test that sufficiently validates its functionality. However, in practice, not every feature has an E2E right now, but, if I am not mistaken, validating the feature still works should be the responsibility of the team owning the feature, if not, I will have to validate myself each and every feature using the tagger, which is impossible due to:
- Lacking context about some (or all) related features
- The large number of related features, especially when the modified internal component is very central. 

So to sum up, when I say here that my change is covered by unit tests and E2E, I mean that I am assuming that impacted features' owners are taking the responsibility of validating that their features still work, and we know that core features are already covered by E2E tests now.

**Review by davidor — APPROVED** [review]

**Review by L3n41c — APPROVED** [review]

**Review by ogaca-dd — APPROVED** [review]
